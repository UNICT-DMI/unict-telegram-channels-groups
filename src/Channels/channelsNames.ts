export const channelsNames: string[] = [
  'dminews',
  'fisicact',
  'dieeinews',
  'deinews',
  'infoDSC',
  'lexunictnews',
  'biogeonews',
  'biometecnews',
  'dspsnews',
  'disumnews',
  'medicina_unict',
  'disfornews',
  'medclinnews',
  'dgfinews',
  'dsfnews',
  'sdslinguenews',
  'dicarnews',
  'ersunews',
  'bio_unict',
  'geo_unict',
  'scienze_ambientali',
  'terapia_unict',
  'Servizio_Sociale_Sociologia_news',
  'lettere_unict',
  'filosofia_unict',
  'IngInfoNews',
  'IngEleNews',
  'IngIndNews',
  'Spotted_DMI',
  'SfottedDIEEI',
];
