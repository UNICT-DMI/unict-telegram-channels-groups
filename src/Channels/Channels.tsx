import React, { useEffect, useState } from 'react';
import channelsNames from './channelsNames';
import ChannelsCards from '../Cards/ChannelsCards';
import Menu from '../Menu/Menu';

interface ChannelEntry {
  title: string;
  link: string;
  description: string;
  pictureURL: string;
  subscribers: number;
}

const API: string = 'https://seminaraluigi.altervista.org/list-telegram-groups/api.telegram.php?';

export default function Channels(): JSX.Element {
  const [channelsArray, setChannelsArray] = useState<ChannelEntry[]>([]);
  const [loading, setLoading] = useState<boolean>(true);
  const [searchInput, setSearchInput] = useState<string>('');

  useEffect(() => {
    const tmpArray: ChannelEntry[] = [];

    async function getData(channelName: string): Promise<void> {
      const newChannelEntry: ChannelEntry = {
        title: '',
        link: '',
        description: '',
        pictureURL: '',
        subscribers: 0,
      };

      const first = fetch(`${API}chat=${channelName}`)
        .then(res => res.json())
        .then(data => {
          newChannelEntry.title = data.result.title;
          newChannelEntry.link = `https://t.me/${channelName}`;
          newChannelEntry.description = data.result.description ? data.result.description : '';
          return fetch(`${API}file=${data.result.photo.big_file_id as string}`);
        })
        .then(res => res.json())
        .then(d => {
          newChannelEntry.pictureURL = `${API}path=${d.result.file_path as string}`;
        });

      const second = fetch(`${API}count=${channelName}`)
        .then(res => res.json())
        .then(data => {
          newChannelEntry.subscribers = data.result;
        });

      await Promise.all([first, second]).then(() => tmpArray.push(newChannelEntry));
    }

    function compare(a: ChannelEntry, b: ChannelEntry): number {
      if (a.subscribers < b.subscribers) return 1;
      if (a.subscribers > b.subscribers) return -1;
      return 0;
    }

    async function initialize() {
      const promises: Promise<void>[] = [];

      channelsNames.forEach(channel => {
        promises.push(getData(channel));
      });

      return Promise.all(promises);
    }

    initialize()
      .then(() => {
        tmpArray.sort(compare);
        setChannelsArray(tmpArray);
        setLoading(false);
      })
      .catch(err => console.log(err));
  }, []);

  return (
    <div>
      <Menu section='channels' setSearchInput={setSearchInput} />
      {loading ? (
        <img src='loading.gif' className='loading' key='loading' alt='loading' />
      ) : (
        <div className='contents-grid'>
          {channelsArray.map(
            (channel, index) =>
              channel.title.toLowerCase().includes(searchInput.toLowerCase()) && (
                <div className='cards' key={channel.title}>
                  <ChannelsCards
                    ranking={index + 1}
                    isSearch={searchInput !== ''}
                    title={channel.title}
                    link={channel.link}
                    description={channel.description}
                    picture={channel.pictureURL}
                    subscribers={channel.subscribers}
                  />
                </div>
              )
          )}
        </div>
      )}
    </div>
  );
}
