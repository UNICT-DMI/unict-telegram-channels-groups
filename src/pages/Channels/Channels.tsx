import { useEffect, useState } from 'react';
import channelsNames from './channelsNames';
import ChannelsCards from '../Cards/ChannelsCards';

interface ChannelProps {
  API: string;
  searchInput: string;
}

interface ChannelEntry {
  title: string;
  link: string;
  description: string;
  pictureURL: string;
  subscribers: number;
}

const Channels: React.FC<ChannelProps> = (props: ChannelProps) => {
  const [channelsArray, setChannelsArray] = useState<ChannelEntry[]>([]);
  const [loading, setLoading] = useState<boolean>(true);

  const promises: Promise<void>[] = [];
  const tmpArray: ChannelEntry[] = [];

  async function getData(channelName: string): Promise<void> {
    const newChannelEntry: ChannelEntry = {
      title: '',
      link: '',
      description: '',
      pictureURL: '',
      subscribers: 0,
    };

    const first = fetch(`${props.API}chat=${channelName}`)
      .then((res) => res.json())
      .then((data) => {
        newChannelEntry.title = data.result.title;
        newChannelEntry.link = `https://t.me/${channelName}`;
        newChannelEntry.description = data.result.description ? data.result.description : '';
        return fetch(`${props.API}file=${data.result.photo.big_file_id as string}`);
      })
      .then((res) => res.json())
      .then((d) => {
        newChannelEntry.pictureURL = `${props.API}path=${d.result.file_path as string}`;
      });

    const second = fetch(`${props.API}count=${channelName}`)
      .then((res) => res.json())
      .then((data) => {
        newChannelEntry.subscribers = data.result;
      });

    await Promise.all([first, second]).then(() => tmpArray.push(newChannelEntry));
  }

  function compare(a: ChannelEntry, b: ChannelEntry): number {
    if (a.subscribers < b.subscribers) return 1;
    if (a.subscribers > b.subscribers) return -1;
    return 0;
  }

  useEffect(() => {
    channelsNames.forEach((channel) => {
      promises.push(getData(channel));
    });

    Promise.all(promises)
      .then(() => {
        tmpArray.sort(compare);
        setChannelsArray(tmpArray);
        setLoading(false);
      })
      .catch((err) => console.log(err));
  }, []);

  return (
    <>
      {loading ? (
        <img src='loading.gif' className='loading' alt='loading' />
      ) : (
        <div className='contents-grid'>
          {channelsArray.map(
            (channel, index) =>
              channel.title.toLowerCase().includes(props.searchInput.toLowerCase()) && (
                <ChannelsCards
                  ranking={index + 1}
                  isSearch={props.searchInput !== ''}
                  title={channel.title}
                  link={channel.link}
                  description={channel.description}
                  picture={channel.pictureURL}
                  subscribers={channel.subscribers}
                />
              )
          )}
        </div>
      )}
    </>
  );
};

export default Channels;
