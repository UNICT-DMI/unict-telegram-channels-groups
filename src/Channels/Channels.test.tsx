import React from 'react';
import { render, cleanup } from '@testing-library/react';
import { Channels } from './Channels';
import { BrowserRouter } from 'react-router-dom';

afterEach(cleanup);

test('should take a snapshot', () => {
  const { asFragment } = render(
    <BrowserRouter>
      <Channels />
    </BrowserRouter>
  );

  expect(asFragment()).toMatchSnapshot();
});
