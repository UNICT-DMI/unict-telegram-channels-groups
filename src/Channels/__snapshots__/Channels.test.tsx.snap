// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should take a snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      class="routing"
    >
      <h1
        class="ranking-title"
      >
        Classifica canali UNICT
      </h1>
      <a
        class="link-to-groups"
        href="/groups"
      >
        Visualizza Gruppi DMI UNICT
      </a>
    </div>
    <input
      class="search-input-field"
      placeholder="Search..."
    />
    <img
      alt="loading"
      class="loading"
      src="loading.gif"
    />
  </div>
</DocumentFragment>
`;
